### PR TITLE
make the buffer in cth_readable_failonly optionaly bound

### PR DESCRIPTION
**Pull request by TheGeorge**
This addresses the issue found in #33

changes:
- config parameter for maximum amount of buffered events
- default to inf, which is buffer all (current behavior)
- change buffer to queue to handle in/out in O(1)
- streamline internal config to make adding the extra config parameter easier

### CODE REVIEW

**Comment by TheGeorge** [comment]
I'm gonna add documentation to the README.md

Currently the existing CTs are failing on master for me, but I'll think of a way to test this

**Comment by ferd** [comment]
Some of the tests are expected to fail, and I check things by hand... otherwise it's hard to write a test suite that relies on other tests' failures and the visibility of their output, though I should probably have replaced this by shelltests long ago.

File: src/cth_readable_failonly.erl [hunk 1]
```diff
@@ -6,7 +6,7 @@
                 handlers=[],
                 named,
                 has_logger}).
--record(eh_state, {buf = [], sasl=false}).
+-record(eh_state, {buf, sasl=false, max_events = inf, stored_events = 0}).
```
Comment (ferd) [inline src/cth_readable_failonly.erl]: you could still use buf = queue:new() here and keep that bit self-initializing.

**Comment by TheGeorge** [comment]
Right, the groups are even called `works` and `fails` :)

File: src/cth_readable_failonly.erl [hunk 2]
```diff
@@ -236,50 +240,69 @@ handle_event(_, S) ->
...
+    S#eh_state{buf=queue:in(Event, Buf), stored_events=StoredEvents + 1};
+buffer_event(Event, S=#eh_state{buf=Buf0}) ->
+    %% bound buffer; buffer filled
+    {_, Buf1} = queue:out(Buf0),
+    S#eh_state{buf=queue:in(Event, Buf1)}.
+
+flush(Buf, Cfg, ShowSASL, SASLType) ->
+    case queue:out(Buf) of
+        {empty, _} -> ok;
+        {{T, Event}, NextBuf} ->
```
Comment (ferd) [inline src/cth_readable_failonly.erl]: The type signature here should make us expect `{{value,  {T, Event}}, NextBuf}`. You're going to get only ignored messages right now.

File: src/cth_readable_failonly.erl [hunk 3]
```diff
@@ -328,10 +351,13 @@ get_sasl_error_logger_type() ->
...
     end.
 
-sasl_ran([]) -> false;
-sasl_ran([{sasl, {_DateTime, {info_report,_,
-            {_,progress, [{application,sasl},{started_at,_}|_]}}}}|_]) -> true;
-sasl_ran([_|T]) -> sasl_ran(T).
+sasl_ran(Buf) ->
+    case queue:out(Buf) of
+        {empty, _} -> false;
+        {{sasl, {_DateTime, {info_report,_,
```
Comment (ferd) [inline src/cth_readable_failonly.erl]: value tagging missing here too

**Comment by TheGeorge** [comment]
I addressed the comments, and have a local test suite, but running that one would require changes to `rebar.config`, which wouldn't be nice. I'll check the suite in, for manual checking anyways.

File: README.md [hunk 4]
```diff
@@ -13,7 +13,9 @@ There are currently the following hooks:
    shell in case of failures. It also provides `cthr:pal/1-4` functions,
    working like `ct:pal/1-4`, but being silenceable by that hook. A parse
    transform exists to automatically convert `ct:pal/1-3` into `cthr:pal/1-3`.
-   Also automatically handles lager.
+   Also automatically handles lager. This hook buffers the IO/logging events,
+   and the buffer size can be limited with the `mmax_events` config option. The
```
Comment (sudhagarc) [inline README.md]:
```suggestion
   and the buffer size can be limited with the `max_events` config option. The
```

**Review by sudhagarc — APPROVED** [review]

**Comment by TheGeorge** [comment]
@ferd can this be merged, or are there any changes you'd like me to do?

**Comment by ferd** [comment]
I wanted to manually run tests and I'm getting issues. The previous (master) version had results like follows:

```
%%% show_logs_SUITE: ......                                                                                                                     
=ERROR REPORT==== 8-Jan-2021::20:41:02.597633 ===                                                                                               
error                                                                                                                                           
                                                                        
=WARNING REPORT==== 8-Jan-2021::20:41:02.597809 ===                                                                                             
warn                                                                    
                                                                                                                                                
=INFO REPORT==== 8-Jan-2021::20:41:02.597943 ===                                                                                                
info                                                                    
                                                                                                                                                
                                                                        
%%% show_logs_SUITE ==> fails.all.error_logger: FAILED                                                                                          
%%% show_logs_SUITE ==> {fail,[{show_logs_SUITE,error_logger,1,                                                                                 
                        [{file,"/home/ferd/code/self/cth_readable/test/show_logs_SUITE.erl"},                                                   
                         {line,48}]},                                                                                                           
       {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1754}]},                                                                            
       {test_server,run_test_case_eval1,6,                                                                                                      
                    [{file,"test_server.erl"},{line,1263}]},            
       {test_server,run_test_case_eval,9,                               
                    [{file,"test_server.erl"},{line,1195}]}]}                                                                          
...
```
Which shows that upon test failure, the logs that were accumulated are displayed.

This current PR instead shows the following:

```
%%% show_logs_SUITE: ......         
%%% show_logs_SUITE ==> fails.all.error_logger: FAILED                                                                                          
%%% show_logs_SUITE ==> {fail,[{show_logs_SUITE,error_logger,1,                                                                                 
                        [{file,"/home/ferd/code/self/cth_readable/test/show_logs_SUITE.erl"},                                                   
                         {line,48}]},
       {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1754}]},    
       {test_server,run_test_case_eval1,6,                              
                    [{file,"test_server.erl"},{line,1263}]},
       {test_server,run_test_case_eval,9,                               
                    [{file,"test_server.erl"},{line,1195}]}]}
...
``` 

All the logs that were supposed to be displayed aren't displayed anymore, and they're eaten up.

This change in behaviour is doing the opposite of what the library should and isn't mergeable.

**Comment by TheGeorge** [comment]
Got, this is I guess a case for automating the tests instead of relying on eyeballing the output. Let me fix the bug and update the PR.

**Comment by TheGeorge** [comment]
I fixed the code and also added a printout of how many events have been dropped:
```
%%% show_logs_SUITE: ......
=ERROR REPORT==== 11-Jan-2021::11:03:04.139807 ===
error

=WARNING REPORT==== 11-Jan-2021::11:03:04.139905 ===
warn

=INFO REPORT==== 11-Jan-2021::11:03:04.140037 ===
info


%%% show_logs_SUITE ==> fails.all.error_logger: FAILED
%%% show_logs_SUITE ==> {fail,[{show_logs_SUITE,error_logger,1,
                        [{file,"/Users/loscher/git/cth_readable/test/show_logs_SUITE.erl"},
                         {line,48}]},
       {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1754}]},
       {test_server,run_test_case_eval1,6,
                    [{file,"test_server.erl"},{line,1263}]},
       {test_server,run_test_case_eval,9,
                    [{file,"test_server.erl"},{line,1195}]}]}
 ```
 
 ```
 (logs are truncated, dropped 9 events)

----------------------------------------------------
2021-01-11 10:07:59.228
Event 9


----------------------------------------------------
2021-01-11 10:07:59.228
Event 10
```

**Pull request merged**